veps/sig-storage/vm-object-graph.md

```diff
@@ -0,0 +1,279 @@
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue created, which links to VEP dir in [kubevirt/enhancements] : https://github.com/kubevirt/enhancements/issues/41
+
+## **Overview**
+
+This is a proposal to include an `Object Graph` API in KubeVirt to represent VM and VMI dependencies and their relationships.
+
+## **Motivation**
+
+As new features continue to be added to KubeVirt, the graph of objects related to VMs (DataVolumes, PersistentVolumeClaims, InstanceTypes, Preferences, Secrets, ConfigMaps, etc.) continues to expand. Identifying all the objects that a VM depends on for tasks like backup, disaster recovery, or migration can be error-prone. We should simplify this process for users and partners by creating an authoritative way to retrieve a structured object graph.
+
+## **Goals**
+
+- Introduce a new Object Graph API to represent the list of resources a VM or VMI depends on.
+- Expose this API as a subresource of VirtualMachines and VirtualMachineInstances.
+- Provide a flexible and extensible data structure.
+- Allow for basic filtering (present or future).
+
+## **Non-Goals**
+
+- Reimplementation of existing VM/VMI specs.
+- Building a generic Kubernetes-wide graph system.
+
+## **User Stories**
+
+1. As a KubeVirt user, I want a clear way to retrieve all VM and VMI-related dependencies.
+2. As a backup partner, I want a way to identify a list of a VM's related objects so I can comprehensively backup and restore everything a VM needs.
+3. As a VM owner, I want to easily define an ACM-discovered application and protect my VM with disaster recovery software.
+4. As a VM owner, I want to migrate my VM from one cluster to another and identify all necessary dependencies for replication.
+5. As a KubeVirt developer, I want a specific place to keep the object graph code updated when I introduce code that changes the relationship of a VM to its dependent objects.
+
+## **Repos**
+
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+- [KubeVirt Velero Plugin](https://github.com/kubevirt/kubevirt-velero-plugin)
+
+## **Design**
+
+### API Endpoints
+
+```
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachines/{name}/objectgraph
+/apis/subresources.kubevirt.io/v1/namespaces/{namespace}/virtualmachineinstances/{name}/objectgraph
+```
+
+### API Examples
+
+#### **1. First option: Hierarchical Representation**
+
+This format visualizes dependencies in a tree structure, showing parent-child relationships between resources.
+
+```go
+// ObjectGraphNode represents an individual resource node in the graph.
+type ObjectGraphNode struct {
+	ObjectReference k8sv1.TypedObjectReference `json:"objectReference"`
+	Labels          map[string]string          `json:"labels,omitempty"`
+	Optional        bool                       `json:"optional"`
+	Children        []ObjectGraphNode          `json:"children,omitempty"`
```

Comment:
I think we should think about having Children being []k8sv1.TypedObjectReference instead. I know KubeVirt doesn't do this but it is theoretically possible for object references to be recursive. obj1 owns obj2 and obj2 own obj1 so I'm not sure we want to have graphs within graphs in the json

Reply:
Your suggestion seems more aligned with the second API example I proposed in the enhancement (flat dependency list). But in a hierarchical format, having full ObjectGraphNodes as children allows for stuff like applying custom labels or filtering subtrees based on those labels. It would feel a bit limiting if only the top-level node could carry those fields, but not its children, when precisely only makes sense for the children to have fields such as `optional`.

Reply:
@mhenriks do you have an example in mind? Is recursion really possible? Since we are describing kubevirt object, we can refuse to put recursive elements in there. Seems a nightmare for deletion if there are those situation

Reply:
@alromeros why are children not included in your flat graph? How does a user know that the datavolume in your example is a child of the VM and not the VMI?

@alicefr Pretty sure there are no recursive references in KubeVirt, coming at this more from a theoretical level and the "flat" representation is more similar to an adjacency list

Reply:
> How does a user know that the datavolume in your example is a child of the VM and not the VMI?

The flat graph is meant to be more utilitarian, it gives you the full list of dependencies for a VM/VMI so users can know what it's exactly needed to recreate it but doesn't go into ownership or relationships. In that context, whether the DV is linked to the VM or VMI might not matter. But yeah that's a good point, I'm starting to think a hierarchical representation is more appropriate.